granary/flickr.py

```diff
@@ -0,0 +1,371 @@
+      },
+      'id': self.tag_uri(username),
+      # numeric_id is our own custom field that always has the source's numeric
+      # user id, if available.
+      'numeric_id': person.get('nsid'),
+      'location': {
+        'displayName': person.get('location', {}).get('_content'),
+      },
+      'username': username,
+      'description': person.get('description', {}).get('_content'),
+    })
+
+    # fetch profile page to get url(s)
+    profile_url = person.get('profileurl', {}).get('_content')
+    if profile_url:
+      try:
+        logging.debug('fetching flickr profile page %s', profile_url)
+        resp = urllib2.urlopen(
+          profile_url, timeout=appengine_config.HTTP_TIMEOUT)
+        profile_json = mf2py.parse(doc=resp, url=profile_url)
+        # personal site is likely the first non-flickr url
+        urls = profile_json.get('rels', {}).get('me', [])
+        obj['urls'] = [{'value': u} for u in urls]
+        obj['url'] = next(
+          (u for u in urls if not u.startswith('https://www.flickr.com/')),
+          None)
+      except urllib2.URLError, e:
+        logging.warning('could not fetch user homepage %s', profile_url)
+
+    return self.postprocess_object(obj)
+
+  def get_comment(self, comment_id, activity_id, activity_author_id=None):
+    """Returns an ActivityStreams comment object.
+
+    Args:
+      comment_id: string comment id
+      activity_id: string activity id, required
+      activity_author_id: string activity author id, ignored
+    """
+    logging.debug('looking for comment %s under photo %s', comment_id, activity_id)
+    resp = self.call_api_method('flickr.photos.comments.getList', {
+      'photo_id': activity_id,
+    })
+    for comment in resp.get('comments', {}).get('comment', []):
+      logging.debug('checking comment id %s', comment.get('id'))
+      # comment id is the in form ###-postid-commentid
+      if (comment.get('id') == comment_id or
+              comment.get('id').split('-')[-1] == comment_id):
+        logging.debug('found comment matching %s', comment_id)
+        return self.comment_to_object(comment, activity_id)
+
+  def photo_to_activity(self, photo):
+    """Convert a Flickr photo to an ActivityStreams object. Takes either
+    data in the expanded form returned by flickr.photos.getInfo or the
+    abbreviated form returned by flickr.people.getPhotos.
+
+    Args:
+      photo: dict response from Flickr
+
+    Returns:
+      dict, an ActivityStreams object
+    """
+    owner = photo.get('owner')
+    owner_id = owner.get('nsid') if isinstance(owner, dict) else owner
+
+    created = photo.get('dates', {}).get('taken') or photo.get('datetaken')
+    published = self.reformat_unix_time(
+      photo.get('dates', {}).get('posted') or photo.get('dateupload'))
+
+    photo_permalink = 'https://www.flickr.com/photos/{}/{}/'.format(
+      owner_id, photo.get('id'))
+
+    title = photo.get('title')
+    if isinstance(title, dict):
```

Comment:
hehe, yeah, it's predictable when it will be a dict and when it will be a string (get a single photo vs. a list of photos), but I have to be cute here to use the same function to handle both.